translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,405 @@
+		return data.TraceData{}
+	}
+
+	resource := ocNodeResourceToInternal(td.Node, td.Resource)
+
+	resourceSpans := &data.ResourceSpans{}
+	resourceSpans.SetResource(resource)
+	resourceSpanList := []*data.ResourceSpans{resourceSpans}
+
+	spanCount := len(td.Spans)
+
+	if spanCount != 0 {
+		// Create slice that holds pointers and another slice that holds structs at once
+		// to avoid individual allocations of structs.
+		content := data.NewSpanSlice(spanCount)
+		ptrs := make([]*data.Span, 0, spanCount)
+
+		for i, ocSpan := range td.Spans {
+			if ocSpan == nil {
+				// Skip nil spans.
+				continue
+			}
+
+			// Point one element in slice of pointers to an element in slice of structs.
+			dest := &content[i]
+
+			ocSpanToInternal(dest, ocSpan)
+
+			if ocSpan.Resource != nil {
+				// Add a separate ResourceSpans item just for this span since it
+				// has a different Resource.
+				separateRS := &data.ResourceSpans{}
+				separateRS.SetResource(ocNodeResourceToInternal(nil, ocSpan.Resource))
+				separateRS.SetSpans([]*data.Span{dest})
+				resourceSpanList = append(resourceSpanList, separateRS)
+			} else {
+				// Otherwise add the span to the first ResourceSpans item.
+				ptrs = append(ptrs, dest)
+			}
+		}
+
+		resourceSpans.SetSpans(ptrs)
+	}
+
+	return data.NewTraceData(resourceSpanList)
+}
+
+func timestampToUnixnano(ts *timestamp.Timestamp) data.TimestampUnixNano {
+	return data.TimestampUnixNano(uint64(internal.TimestampToTime(ts).UnixNano()))
+}
+
+func ocSpanToInternal(dest *data.Span, src *octrace.Span) {
+	events, droppedEventCount := ocEventsToInternal(src.TimeEvents)
+	links, droppedLinkCount := ocLinksToInternal(src.Links)
+
+	dest.SetTraceID(data.TraceIDFromBytes(src.TraceId))
+	dest.SetSpanID(data.SpanIDFromBytes(src.SpanId))
+	dest.SetTraceState(ocTraceStateToInternal(src.Tracestate))
+	dest.SetParentSpanID(data.SpanIDFromBytes(src.ParentSpanId))
+	dest.SetName(truncableStringToStr(src.Name))
+	dest.SetKind(ocSpanKindToInternal(src.Kind, src.Attributes))
+	dest.SetStartTime(timestampToUnixnano(src.StartTime))
+	dest.SetEndTime(timestampToUnixnano(src.EndTime))
+	dest.SetAttributes(ocAttrsToInternal(src.Attributes))
+	dest.SetEvents(events)
+	dest.SetDroppedEventsCount(droppedEventCount)
+	dest.SetLinks(links)
+	dest.SetDroppedLinksCount(droppedLinkCount)
+	dest.SetStatus(ocStatusToInternal(src.Status))
+}
+
+func ocStatusToInternal(ocStatus *octrace.Status) data.SpanStatus {
+	if ocStatus == nil {
+		return data.SpanStatus{}
+	}
+	return data.NewSpanStatus(data.StatusCode(ocStatus.Code), ocStatus.Message)
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToInternal(ocTracestate *octrace.Span_Tracestate) data.TraceState {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
```

Comment:
Might be redundant creating a slice here, could be replaced with
```
sb.WriteString(entry.Key)
sb.WriteString("=")
sb.WriteString(entry.Value)
```

Reply:
done